File: medium/Hash-Table/first-unique-character-in-a-string/first_uniq_char.py
class Solution:
    def firstUniqChar(self, s: str) -> int:
        """
        Find the first non-repeating character in a string and return its index.
        Approach: Hash Map (frequency counting)
        Time Complexity: O(n)
        Space Complexity: O(1) (only 26 lowercase letters)
        """
        # Count frequency of each character
        freq = {}
        for c in s:
            freq[c] = freq.get(c, 0) + 1

        # Find the first character with frequency 1
        for idx, c in enumerate(s):
            if freq[c] == 1:
                return idx

        # No unique character exists
        return -1

File: medium/Hash-Table/first-unique-character-in-a-string/first_uniq_char.cpp
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    int firstUniqChar(string s) {
        /*
         * Find the first non-repeating character and return its index.
         * Approach: Hash map for frequency counting
         * Time Complexity: O(n)
         * Space Complexity: O(1)
         */
        unordered_map<char, int> count;

        // Count each character
        for (char c : s) {
            count[c]++;
        }

        // Find first unique one
        for (int i = 0; i < s.size(); ++i) {
            if (count[s[i]] == 1) {
                return i;
            }
        }

        return -1;
    }
};
